fabric-recipe-api-v1/src/main/java/net/fabricmc/fabric/impl/recipe/ingredient/CustomIngredientImpl.java

```diff
@@ -0,0 +1,143 @@
+		}
+	}
+
+	@Nullable
+	public static CustomIngredientSerializer<?> getSerializer(Identifier identifier) {
+		Objects.requireNonNull(identifier, "Identifier may not be null.");
+
+		return REGISTERED_SERIALIZERS.get(identifier);
+	}
+
+	static {
+		CustomIngredientSerializer.register(AndIngredient.SERIALIZER);
+		CustomIngredientSerializer.register(OrIngredient.SERIALIZER);
+		CustomIngredientSerializer.register(DifferenceIngredient.SERIALIZER);
+		CustomIngredientSerializer.register(NbtIngredient.SERIALIZER);
+	}
+
+	// Actual custom ingredient logic
+
+	private final CustomIngredient customIngredient;
+
+	public CustomIngredientImpl(CustomIngredient customIngredient) {
+		super(Stream.empty());
+
+		this.customIngredient = customIngredient;
+	}
+
+	@Override
+	public CustomIngredient getCustomIngredient() {
+		return customIngredient;
+	}
+
+	@Override
+	public boolean requiresTesting() {
+		return customIngredient.requiresTesting();
+	}
+
+	@Override
+	public ItemStack[] getMatchingStacks() {
+		if (this.matchingStacks == null) {
+			this.matchingStacks = customIngredient.getMatchingStacks();
+		}
+
+		return this.matchingStacks;
+	}
+
+	@Override
+	public boolean test(@Nullable ItemStack stack) {
+		return stack != null && customIngredient.test(stack);
+	}
+
+	@Override
+	public void write(PacketByteBuf buf) {
+		// Can be null if we're not writing a packet from the PacketEncoder; in that case, always write the full ingredient.
+		// Chances are this is a mod's doing and the client has the Ingredient API with the relevant ingredients.
+		Set<Identifier> supportedIngredients = CustomIngredientSync.CURRENT_SUPPORTED_INGREDIENTS.get();
+
+		if (supportedIngredients != null && !supportedIngredients.contains(customIngredient.getSerializer().getIdentifier())) {
+			// The client doesn't support this custom ingredient, so we send the matching stacks as a regular ingredient.
+			// Conveniently, this is exactly what the super call does.
+			super.write(buf);
+		} else {
+			// The client supports this custom ingredient, so we send it as a custom ingredient.
+			buf.writeVarInt(PACKET_MARKER);
+			buf.writeIdentifier(customIngredient.getSerializer().getIdentifier());
+			customIngredient.getSerializer().write(buf, coerceIngredient());
+		}
+	}
+
+	@Override
+	public JsonElement toJson() {
+		JsonObject json = new JsonObject();
+		json.addProperty(TYPE_KEY, customIngredient.getSerializer().getIdentifier().toString());
+		customIngredient.getSerializer().write(json, coerceIngredient());
+		return json;
+	}
+
+	@Override
+	public boolean isEmpty() {
+		// We don't want to resolve the matching stacks,
+		// as this might cause the ingredient to use outdated tags when it's done too early.
+		// So we just return false when the matching stacks haven't been resolved yet (i.e. when the field is null).
+		// TODO: this is a bit hacky, can we not do better?
```

Comment:
When is this called to early? Would hard crashing be better?

Reply:
![image](https://user-images.githubusercontent.com/13494793/208391470-89dc96c5-8fea-4ac0-a73c-61895ab4389d.png)

For example when used in a shapeless recipe. Vanilla tries to filter out empty ingredients for some reason, but tags haven't been bound yet, so the ingredient should not be resolved.

I don't know if we need to do anything about this, and in my testing this was never really an issue, so I'd suggest we leave it as-is for now. I removed the TODO.